res/controllers/common-hid-packet-parser.js

```diff
@@ -1,1549 +1,1800 @@
+ *
+ * @callback scalingCallback
+ * @param {string} group Defines the group name for the field. The group can be any string, but if it matches a valid Mixxx control group name, it is possible to map a field to a control or output without any additional code.
+ * @param {string} name Is the name of the control for the field. The name can be any string, but if it matches a valid Mixxx control name in the group defined for field, the system attempts to attach it directly to the correct field. Together group and name form the ID of the field (group.name)
+ * @param {number} value Value to be scaled
+ * @returns {number} Scaled value
+ */
+/**
+ * @typedef packetField
+ * @type {object}
+ * @property {HIDPacket} packet
+ * @property {string} id Group and control name separated by a dot
+ * @property {string} group
+ * @property {string} name
+ * @property {string} mapped_group
+ * @property {string} mapped_name
+ * @property {object} pack Control packing format for unpack(), one of b/B, h/H, i/I
+ * @property {number} offset
+ * @property {number} end_offset
+ * @property {number} bitmask
+ * @property {boolean} isEncoder
+ * @property {controlCallback} callback
+ * @property {boolean} soft_takeover
+ * @property {boolean} ignored
+ * @property {controlCallback} auto_repeat
+ * @property {number} auto_repeat_interval
+ * @property {number} min
+ * @property {number} max
+ * @property {string} type
+ * @property {HIDBitVector|boolean|number} value
+ * @property {number} delta
+ * @property {number} mindelta
+ * @property {number} toggle
+ */
 
-/** HID Bit Vector Class
+/**
+ * HID Bit Vector Class
  *
  * Collection of bits in one parsed packet field. These objects are
  * created by HIDPacket addControl and addOutput and should not be
- * created manually. */
-var HIDBitVector = function() {
-    this.size = 0;
-    this.bits = {};
-};
-
-/** Return bit offset based on bitmask */
-HIDBitVector.prototype.getOffset = function(bitmask) {
-    for (var i = 0; i < 32; i++)
-        if ((1 & bitmask >> i) !== 0)
-            return i;
-    return 0;
-};
+ * created manually.
+ */
+class HIDBitVector {
+    constructor() {
+        this.size = 0;
+        this.bits = {};
+    }
+    /**
+     * Get the index of the least significant bit that is 1 in `bitmask`
+     *
+     * @param {number} bitmask A bitwise mask of up to 32 bit. All bits set to'1' in this mask are considered.
+     * @returns {number} Index of the least significant bit that is 1 in `bitmask`
+     */
+    getOffset(bitmask) {
+        for (let i = 0; i < 32; i++) {
+            if ((1 & bitmask >> i) !== 0) { return i; }
+        }
+        return 0;
```

Comment:
With the typescript declarations from #4759 it's really easy to see if a function is available in Mixxx 2.4. The Editor would mark it immediately as error, as you can see this is not the case here:
![grafik](https://user-images.githubusercontent.com/64457745/178114145-2cf64619-2c16-4ee1-9a87-9b7a87a59f81.png)
I also tested this implementation with my mapping and it works flawless.

Reply:
Great. I also tested it on JSBench and its 80% faster on average. Even though that is obviously not the same javascript environment, I'm pretty sure it won't be slower than the current solution.

Reply:
Any performance improvement helps in such library code. Thank you!
My profiling with Trace showed, that the function where the most time is wasted in this library, is getOutputField - an often called function with many nested loops inside.

Reply:
Yeah makes sense... I'm not familiar enough with the code to see any obvious optimization opportunity. I guess it would be nice if you could add a note to the JSDoc to document that that method is expensive...

Reply:
Will do! I plan to address this in a follow-up PR - I think some kind of looktable is needed here.